test/vms.destroy.test.js

```diff
@@ -0,0 +1,532 @@
+                'one we just tried to destroy');
+            t.done();
+        });
+};
+
+exports.put_on_destroying_vm_with_non_destroyed_state_must_fail = function (t) {
+    async.each(common.VALID_VM_STATES, function (vmState, next) {
+        var vmInNewState = clone(vmObject);
+        vmInNewState.state = vmState;
+        client.put(vmLocation, vmInNewState, function (err, req, res, body) {
+            if (vmInNewState.state !== vmObject.state) {
+                // Setting the state of a VM that is being destroyed
+                // to another state that is *not* 'destroyed' should result
+                // in an error.
+                t.equal(res.statusCode, 409);
+                var expectedError = {
+                    code: 'VMBeingDestroyed',
+                    message:
+                        'Invalid operation while this VM is being destroyed'
+                };
+                t.deepEqual(body, expectedError);
+            } else {
+                // However, setting the state of a VM that is being destroyed
+                // to the *same* state is valid.
+                t.ifError(err);
+                t.equal(res.statusCode, 200);
+            }
+
+            return next();
+        });
+    }, function allDone(err) {
+        t.ifError(err);
+        t.done();
+    });
+};
+
+exports.put_destroying_vm_non_destroyed_zone_state_must_fail = function (t) {
+    var nonDestroyedZoneStates =
+        common.VALID_VM_ZONE_STATES.filter(function (state) {
+            return state !== 'destroyed';
+        });
+
+    async.each(nonDestroyedZoneStates, function (vmZoneState, next) {
+        var vmInNewZoneState = clone(vmObject);
+        vmInNewZoneState.zone_state = vmZoneState;
+        client.put(vmLocation, vmInNewZoneState,
+            function (err, req, res, body) {
+                if (vmInNewZoneState.zone_state !== vmObject.zone_state) {
+                    // Setting the zone_state of a VM that is being destroyed
+                    // to another state that is *not* 'destroyed' should result
+                    // in an error.
+                    t.equal(res.statusCode, 409);
+                    var expectedError = {
+                        code: 'VMBeingDestroyed',
+                        message:
+                            'Invalid operation while this VM is being destroyed'
+                    };
+                    t.deepEqual(body, expectedError);
+                } else {
+                    // However, setting the zone_state of a VM that is being
+                    // destroyed to the *same* zone_state is valid.
+                    t.ifError(err);
+                    t.equal(res.statusCode, 200);
+                }
+
+                return next();
+            });
+    }, function allDone(err) {
+        t.ifError(err);
+        t.done();
+    });
+};
+
+exports.post_on_destroying_vm_must_fail = function (t) {
```

Comment:
There's already the [`destroy_destructible_vm`](https://github.com/joyent/sdc-vmapi/pull/11/files#diff-bf4e451cca98fd1a17105a1393a19e4cR408) test that deletes the test VM when it's in the `destroying` state and its `indestructible_zoneroot` property has been set to `false`.

Did you have another test in mind?